frontend/src/v5/ui/components/viewer/drawingViewer/viewerLayer2D/viewerLayer2D.component.tsx

```diff
@@ -46,60 +69,66 @@ export const ViewerLayer2D = ({ viewBox, active, value, onChange }: ViewerLayer2
 		height: `${viewBox.height}px`,
 	};
 
-	// This returns the offset of the cursor from the top-left corner
+	const handleMouseDown = () => previousViewBox.current = viewBox;
+
 	const getCursorOffset = (e) => {
 		const rect = e.target.getBoundingClientRect();
 		const offsetX = e.clientX - rect.left;
 		const offsetY = e.clientY - rect.top;
 		return [offsetX, offsetY].map((point) => point / viewBox.scale) as Coord2D;
 	};
 
-	const handleMouseDown = () => previousViewBox.current = viewBox;
+	const handleMouseUp = useCallback((e) => {
+		let mousePosition = getCursorOffset(e);
 
-	const handleMouseUp = () => {
 		// check if mouse up was fired after dragging or if it was an actual click
 		if (!isEqual(viewBox, previousViewBox.current)) return;
 
-		if (offsetEnd || (!offsetEnd && !offsetStart)) {
-			setOffsetEnd(null);
-			setOffsetStart(mousePosition);
-			onChange(EMPTY_VECTOR);
-		} else if (!isEqual(offsetStart, mousePosition)) {
-			setOffsetEnd(mousePosition);
-			onChange?.([offsetStart, mousePosition]);
+		if (snapping) {
+			const radius = 10 / viewBox.scale;
+			const res = snap(mousePosition, snapHandler, radius);
+			mousePosition = res.mousePos;
+
+			if (snapType !== res.snapType) {
+				setSnapType(res.snapType);
+			}
 		}
-	};
 
-	const handleMouseMove = (e) => {
-		setMousePosition(getCursorOffset(e));
-	};
+		DrawingViewerService.emitMouseClickEvent(mousePosition);
+	}, [viewBox]);
 
-	const resetArrow = () => {
-		setOffsetStart(null);
-		setOffsetEnd(null);
-	};
+	const handleMouseMove = useCallback((e) => {
+		let mousePos = getCursorOffset(e);
 
-	useEffect(() => {
-		if (!active && !offsetEnd) {
-			resetArrow();
+		if (snapping) {
+			const radius = 10 / viewBox.scale;
+			const res = snap(mousePos, snapHandler, radius);
+			mousePos = res.mousePos;
+
+			if (snapType !== res.snapType) {
+				setSnapType(res.snapType);
+			}
```

Comment:
The code above is the same as the previous useEffect
![image](https://github.com/user-attachments/assets/0895f112-dda7-4dae-8b49-1d8b568988cb)
(I moved `code A` up to better show the differences)